### PR TITLE
Add GOVUK Frontend

### PR DESCRIPTION
**Pull request by tombye**
Replaces the base layout for all pages, brings in core Sass, for layout & typography, and swaps out the button to use a component.

https://www.pivotaltracker.com/story/show/169475447

Mostly copies the work on the admin app: https://github.com/alphagov/notifications-admin/pull/3158

### Before

download_old

### After

download_new

### Before

document_old

### After

document_new

### Before
profile_overview_old

### After
profile_overview_new

### CODE REVIEW

File: app/templates/error/401.html [hunk 1]
```diff
@@ -1,16 +1,18 @@
...
-      </p>
-      <p>
-        To report the problem, email <a href="mailto:notify-support@digital.cabinet-office.gov.uk">notify-support@digital.cabinet-office.gov.uk</a>.
-      </p>
+      <div class="govuk-body">
+          <p>
+            You cannot download your file at the moment.
+          </p>
+          <p>
+            To report the problem, email <a href="mailto:notify-support@digital.cabinet-office.gov.uk">notify-support@digital.cabinet-office.gov.uk</a>.
```
Comment (idavidmcdonald) [inline app/templates/error/401.html]: Should this have the `govuk-link` class too?

File: app/templates/error/403.html [hunk 2]
```diff
@@ -1,16 +1,18 @@
...
-      </p>
-      <p>
-        To report the problem, email <a href="mailto:notify-support@digital.cabinet-office.gov.uk">notify-support@digital.cabinet-office.gov.uk</a>.
-      </p>
+      <div class="govuk-body">
+          <p>
+            You cannot download your file at the moment.
+          </p>
+          <p>
+            To report the problem, email <a href="mailto:notify-support@digital.cabinet-office.gov.uk">notify-support@digital.cabinet-office.gov.uk</a>.
```
Comment (idavidmcdonald) [inline app/templates/error/403.html]: Should this have the `govuk-link` class too?

File: app/assets/javascripts/modules/all.mjs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+// a bit like `app/__init__` in the Flask app.
+//
+// When processed by a bundler, this is turned into a Immediately Invoked Function Expression (IIFE)
+// The IIFE format allows it to run in browsers that don't support JS Modules.
+//
+// Exported items will be added to the window.GOVUK namespace.
+// For example, `export { Frontend }` will assign `Frontend` to `window.Frontend`
+import Button from 'govuk-frontend/govuk/components/button/button';
+
+// Copy of the initAll function from https://github.com/alphagov/govuk-frontend/blob/v2.13.0/src/all.js
```
Comment (idavidmcdonald) [inline app/assets/javascripts/modules/all.mjs]: Version number still old

File: app/templates/document_download_template.html [hunk 4]
```diff
@@ -10,10 +10,18 @@
 {% endblock %}
 
 {% block head %}
-  <link rel="stylesheet" media="screen" href="{{ asset_url('stylesheets/main.css') }}" />
+  <!--[if !IE 8]><!-->
+    <link rel="stylesheet" media="screen" href="{{ asset_url('stylesheets/main.css') }}" />
+  <!--<![endif]-->
+  <!--[if IE 8]>
+    <link rel="stylesheet" href="{{ asset_url('stylesheets/main-ie8.css') }}">
```
Comment (idavidmcdonald) [inline app/templates/document_download_template.html]: No `media=screen` attribute here intentional?

**Comment by idavidmcdonald** [comment]
image

Some differences in footer colour and colour of text between our current staging and running this locally. I mean, it looks OK.

Maybe more worth comparing with the admin app though which also has colour and font differences. (I assume the lack of Privacy/Cookies and Built by links are also fine).

image

File: app/assets/stylesheets/main.scss [hunk 5]
```diff
@@ -25,6 +25,24 @@ $path: '/static/images/';
 @import 'elements/layout';
 @import 'elements/panels';
 
+// Dependencies from GOV.UK Frontend
+// https://github.com/alphagov/govuk-frontend/tree/v2.13.0
```
[the remark this answers is not in the document]
Comment (tombye) [reply app/assets/stylesheets/main.scss]: Yes! I thought I got then all but missed this, will fix.

File: app/assets/stylesheets/main.scss [hunk 6]
```diff
@@ -25,6 +25,24 @@ $path: '/static/images/';
...
+// set asset URL root to match that of application
+$govuk-assets-path: "/static/";
+
+@import "settings/all";
+@import "tools/all";
+@import "helpers/all";
+
+@import "core/all";
+@import "objects/all";
+
```
[the remark this answers is not in the document]
Comment (tombye) [reply app/assets/stylesheets/main.scss]: Exactly that. I commented it in the admin app but missed that out here. (see https://github.com/alphagov/notifications-admin/blob/master/app/assets/stylesheets/govuk-frontend/_all.scss#L1)

Will update with the same comments.

File: app/templates/error/401.html (hunk 1 above)
Comment (tombye) [reply app/templates/error/401.html]: Yes! Missed that one, will add a fix.

File: app/templates/document_download_template.html (hunk 4 above)
Comment (tombye) [reply app/templates/document_download_template.html]: I copied it from the [GOVUK Frontend docs](https://github.com/alphagov/govuk-frontend/blob/master/docs/installation/supporting-internet-explorer-8.md#including-the-ie8-stylesheet-in-your-project) but yes, it should match our other tags.

File: app/templates/error/403.html (hunk 2 above)
Comment (tombye) [reply app/templates/error/403.html]: Yes! Missed that one, will add a fix.

**Comment by tombye** [comment]
> image
> 
> Some differences in footer colour and colour of text between our current staging and running this locally. I mean, it looks OK.
> 
> Maybe more worth comparing with the admin app though which also has colour and font differences. (I assume the lack of Privacy/Cookies and Built by links are also fine).
> 
> image

The differences in colour are intentional and came in as part of the work described in [the design system team's blog post about changes to increase accessibility](https://designnotes.blog.gov.uk/2019/07/29/weve-made-the-gov-uk-design-system-more-accessible/).

This app doesn't set cookies so I think doesn't need to link to the cookies page. I'm not as sure about privacy but it isn't in the existing version. @yahoopete or @quis might know.

**Comment by idavidmcdonald** [comment]
@quis tells me the differences in footer is all fine due to not upgrading admin either yet to this point and a switch in the colour palette.

**Comment by yahoopete** [comment]
As we are the data processor and not the data controller, the privacy aspects are not for us to state, so no need for them here. The ones on the admin app are appropriate as we are the data controller of the user data.

File: app/assets/javascripts/modules/all.mjs (hunk 3 above)
Comment (tombye) [reply app/assets/javascripts/modules/all.mjs]: This code is now in `app/assets/javascripts/govuk-frontend-local.mjs` so I've updated it there.

File: requirements.txt [hunk 7]
```diff
@@ -33,28 +34,28 @@ future==0.18.2
...
 phonenumbers==8.10.13
-pyasn1==0.4.7
+pyasn1==0.4.8
 PyJWT==1.7.1
 PyPDF2==1.26.0
 python-dateutil==2.8.1
 python-json-logger==0.1.11
 pytz==2018.9
-PyYAML==5.1
+PyYAML==3.13
```
[the remark this answers is not in the document]
Comment (tombye) [reply requirements.txt]: Looks like it was brought in by the version of `notifications-utils` being quite old. I've bumped it (in https://github.com/alphagov/document-download-frontend/pull/53/commits/d31006d8b4c1ce01aac5a17f826dfa697f497d45) to the latest version which brings it up to a version without the vulnerability.

**Comment by tombye** [comment]
> Good work.
> 
> If it's easy for you to do so, you may want to try and run the functional tests against this locally before merging if you haven't already.

That's a good point, will do 👍

File: app/templates/error/404.html [hunk 8]
```diff
@@ -1,20 +1,22 @@
...
-          If you selected a link in an email, contact the sender to tell them you cannot download your file.
-        </p>
-        <p>
-          If you typed the web address, check it is correct.
-        </p>
-        <p>
-          If you pasted the web address, check you copied the entire address.
-        </p>
+        <div class="govuk-body">
+          <p>
```
Comment (nickcolley) [inline app/templates/error/404.html]: 'govuk-body' classes should be on each paragraph element

File: app/templates/views/download.html [hunk 9]
```diff
@@ -1,22 +1,30 @@
...
-    {% else %}
-      call {{ service_contact_info }}.
-    {% endif %}
-  </p>
+    <p>
+      If you have any questions,
+      {% if contact_info_type == "link" %}
+        <a href={{ service_contact_info }} class="govuk-link"> contact {{ service_name }}</a>.
+      {% elif contact_info_type == "email" %}
+        email <a href={{ "mailto:" + service_contact_info }} class="govuk-link">{{service_contact_info}}</a>.
```
Comment (nickcolley) [inline app/templates/views/download.html]: Should these hrefs values have string quotes around them?

File: app/templates/error/404.html (hunk 8 above)
Comment (tombye) [reply app/templates/error/404.html]: I didn't know that was a rule. Is it wrong to apply it to a block of paragraphs?

Comment (nickcolley) [reply app/templates/error/404.html]: I think the spacing will not be correct.

File: app/templates/views/download.html (hunk 9 above)
Comment (tombye) [reply app/templates/views/download.html]: Yes. Will raise a PR for that later.

File: app/templates/error/404.html (hunk 8 above)
Comment (tombye) [reply app/templates/error/404.html]: @quis fixed this in https://github.com/alphagov/document-download-frontend/pull/54/commits/d790b7d86047117dff51752931aab02f22cc6af7 so marking as resolved.

File: app/templates/views/download.html (hunk 9 above)
Comment (tombye) [reply app/templates/views/download.html]: Added in https://github.com/alphagov/document-download-frontend/pull/55 so marking this as resolved.

**Pull request merged**